ledger/ledgercore/statedelta.go

```diff
@@ -69,3 +69,80 @@ type StateDelta struct {
 	// previous block timestamp
 	PrevTimestamp int64
 }
+
+// AccountsDelta stores orderd accounts and allow fast lookup by address
+type AccountsDelta struct {
+	accts      []basics.BalanceRecord
+	acctsCache map[basics.Address]int
+}
+
+// MakeStateDelta creates a new instance of StateDelta
+func MakeStateDelta(hdr *bookkeeping.BlockHeader, prevTimestamp int64) StateDelta {
+	return StateDelta{
+		Accts:         AccountsDelta{},
+		Txids:         make(map[transactions.Txid]basics.Round),
+		Txleases:      make(map[Txlease]basics.Round),
+		Creatables:    make(map[basics.CreatableIndex]ModifiedCreatable),
+		Hdr:           hdr,
+		PrevTimestamp: prevTimestamp,
+	}
+}
+
+// Get lookups AccountData by address
+func (ad *AccountsDelta) Get(addr basics.Address) (basics.AccountData, bool) {
+	idx, ok := ad.acctsCache[addr]
+	if !ok {
+		return basics.AccountData{}, false
+	}
+	return ad.accts[idx].AccountData, true
+}
+
+// ModifiedAccounts returns list of addresses of modified accounts
+func (ad *AccountsDelta) ModifiedAccounts() []basics.Address {
+	result := make([]basics.Address, len(ad.accts))
+	for i := 0; i < len(ad.accts); i++ {
+		result[i] = ad.accts[i].Addr
+	}
+	return result
+}
+
+// MergeAccounts applies other accounts into this StateDelta accounts
+func (ad *AccountsDelta) MergeAccounts(other *AccountsDelta) {
+	for new := range other.accts {
+		ad.addAccount(other.accts[new])
+	}
+}
+
+// Len returns number of stored accounts
+func (ad *AccountsDelta) Len() int {
+	return len(ad.accts)
+}
+
+// GetByIdx returns address and AccountData
+// It does NOT check boundaries.
+func (ad *AccountsDelta) GetByIdx(i int) (basics.Address, basics.AccountData) {
```

Comment:
nit: it looks that the two locations that do need this function would benefit more from having a direct access to the underlying slice.

Reply:
you mean returning pointers or `GetAddrIdx`/`GetDataIdx` methods?

Reply:
Return the inner slice instead of tunnelling each iteration through a getter.